beacon-chain/rpc/beaconv1/state.go

```diff
@@ -2,29 +2,166 @@ package beaconv1
+	curJustChkpt := requestedState.CurrentJustifiedCheckpoint()
+	finalizedChkpt := requestedState.FinalizedCheckpoint()
+	resp := &ethpb.StateFinalityCheckpointResponse{
+		PreviousJustified: &ethpb.Checkpoint{
+			Epoch: prevJustChkpt.Epoch,
+			Root:  prevJustChkpt.Root,
+		},
+		CurrentJustified: &ethpb.Checkpoint{
+			Epoch: curJustChkpt.Epoch,
+			Root:  curJustChkpt.Root,
+		},
+		Finalized: &ethpb.Checkpoint{
+			Epoch: finalizedChkpt.Epoch,
+			Root:  finalizedChkpt.Root,
+		},
+	}
+	return resp, nil
+}
+
+func (bs *Server) getState(ctx context.Context, stateId []byte) (*state.BeaconState, error) {
+	var requestedState *state.BeaconState
+	var err error
+	switch string(stateId) {
+	case "head":
+		requestedState, err = bs.ChainInfoFetcher.HeadState(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get head state")
+		}
+	case "genesis":
+		requestedState, err = bs.StateGen.StateByRoot(ctx, params.BeaconConfig().ZeroHash)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get genesis checkpoint")
+		}
+	case "finalized":
+		finalizedCheckpoint := bs.ChainInfoFetcher.FinalizedCheckpt()
+		requestedState, err = bs.StateGen.StateByRoot(ctx, bytesutil.ToBytes32(finalizedCheckpoint.Root))
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get finalized checkpoint")
+		}
+	case "justified":
+		justifiedCheckpoint := bs.ChainInfoFetcher.CurrentJustifiedCheckpt()
+		requestedState, err = bs.StateGen.StateByRoot(ctx, bytesutil.ToBytes32(justifiedCheckpoint.Root))
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get justified checkpoint")
+		}
+	default:
+		if len(stateId) == 32 {
+			requestedState, err = bs.StateGen.StateByRoot(ctx, bytesutil.ToBytes32(stateId))
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get state")
+			}
+		} else {
+			requestedSlot, err := strconv.ParseUint(string(stateId), 10, 64)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Cannot parse slot from input %#x: %v", stateId, err)
+			}
+			currentSlot := bs.ChainInfoFetcher.HeadSlot()
+			if requestedSlot > currentSlot {
+				return nil, status.Errorf(
+					codes.InvalidArgument,
+					"Cannot retrieve information about a slot in the future, current slot %d, requesting %d",
+					currentSlot,
+					requestedSlot,
+				)
+			}
+			requestedEpoch := helpers.SlotToEpoch(requestedSlot)
+			currentEpoch := helpers.SlotToEpoch(currentSlot)
+			if requestedEpoch > currentEpoch {
+				return nil, status.Errorf(
+					codes.InvalidArgument,
+					"Cannot retrieve information about an epoch in the future, current epoch %d, requesting %d",
+					currentEpoch,
+					requestedEpoch,
+				)
+			}
+
+			requestedState, err = bs.StateGen.StateBySlot(ctx, requestedSlot)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get state")
+			}
+		}
+	}
+	if requestedState == nil {
+		return nil, errors.New("could not find state with state id")
+	}
```

Comment:
Do not do this. Otherwise the caller will not know if it should return a 404 or 500. 

if state == nil and err == nil, that's a 404. 
if err != nil, that's a 500.

Reply:
Also, deepsource is showing this is untested anyway.